Add specialized lerp kernel to accelerate `make_tensor`

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84846
* #85473
* #84845
* #84844

`make_tensor` used several times for every single opinfo test, so small gains in
performace will result in significant improvements in test completion time.
Aside from the random number generation, the main limitation for floating an
complex dtypes is the scaling operation from [0, 1] to [low, high]. This
operation is a linear interpolation (`lerp`) but calling `torch.lerp` is
actually not faster because the kernel isn't optimised for scalar `low` and
`high` values. Particularly on CUDA, this means extra kernel launches and device
synchronization to copy the values into tensors.

This adds a new `_lerp_scalar` kernel for use in `make_tensor` and also reduces
the number of seperate kernels launched for complex tensors, since for the small
tensors used in tests it is often the kernel launch overhead that is the
limiting factor.

Here are benchmark results for `make_tensor` with small and medium sized tensors.

| Device | dtype     | size | Master (us) | This PR (us) | Speedup |
|--------|-----------|------|-------------|--------------|---------|
| cpu    | float32   | 8    | 19.6        | 8.61         | 2.3     |
|        |           | 4096 | 34.9        | 23.3         | 1.5     |
|        | complex64 | 8    | 37.4        | 13.4         | 2.8     |
|        |           | 4096 | 70.3        | 46.0         | 1.5     |
| cuda   | float32   | 8    | 38.7        | 28.2         | 1.4     |
|        |           | 4096 | 36.2        | 18.3         | 1.4     |
|        | complex64 | 8    | 74.9        | 32.3         | 2.3     |
|        |           | 4096 | 74.1        | 32.6         | 2.3     |

peterbell10 committed:
Add specialized lerp kernel to accelerate `make_tensor`

`make_tensor` used several times for every single opinfo test, so small gains in
performace will result in significant improvements in test completion time.
Aside from the random number generation, the main limitation for floating an
complex dtypes is the scaling operation from [0, 1] to [low, high]. This
operation is a linear interpolation (`lerp`) but calling `torch.lerp` is
actually not faster because the kernel isn't optimised for scalar `low` and
`high` values. Particularly on CUDA, this means extra kernel launches and device
synchronization to copy the values into tensors.

This adds a new `_lerp_scalar` kernel for use in `make_tensor` and also reduces
the number of seperate kernels launched for complex tensors, since for the small
tensors used in tests it is often the kernel launch overhead that is the
limiting factor.

Here are benchmark results for `make_tensor` with small and medium sized tensors.

| Device | dtype     | size | Master (us) | This PR (us) | Speedup |
|--------|-----------|------|-------------|--------------|---------|
| cpu    | float32   | 8    | 19.6        | 8.61         | 2.3     |
|        |           | 4096 | 34.9        | 23.3         | 1.5     |
|        | complex64 | 8    | 37.4        | 10.1         | 2.7     |
|        |           | 4096 | 70.3        | 39.2         | 1.8     |
| cuda   | float32   | 8    | 38.7        | 28.2         | 1.4     |
|        |           | 4096 | 36.2        | 18.3         | 1.4     |
|        | complex64 | 8    | 74.9        | 20.6         | 3.6     |
|        |           | 4096 | 74.1        | 20.8         | 3.6     |

[ghstack-poisoned]

peterbell10 committed:
Update on "Add specialized lerp kernel to accelerate `make_tensor`"


`make_tensor` used several times for every single opinfo test, so small gains in
performace will result in significant improvements in test completion time.
Aside from the random number generation, the main limitation for floating an
complex dtypes is the scaling operation from [0, 1] to [low, high]. This
operation is a linear interpolation (`lerp`) but calling `torch.lerp` is
actually not faster because the kernel isn't optimised for scalar `low` and
`high` values. Particularly on CUDA, this means extra kernel launches and device
synchronization to copy the values into tensors.

This adds a new `_lerp_scalar` kernel for use in `make_tensor` and also reduces
the number of seperate kernels launched for complex tensors, since for the small
tensors used in tests it is often the kernel launch overhead that is the
limiting factor.

Here are benchmark results for `make_tensor` with small and medium sized tensors.

| Device | dtype     | size | Master (us) | This PR (us) | Speedup |
|--------|-----------|------|-------------|--------------|---------|
| cpu    | float32   | 8    | 19.6        | 8.61         | 2.3     |
|        |           | 4096 | 34.9        | 23.3         | 1.5     |
|        | complex64 | 8    | 37.4        | 10.1         | 2.7     |
|        |           | 4096 | 70.3        | 39.2         | 1.8     |
| cuda   | float32   | 8    | 38.7        | 28.2         | 1.4     |
|        |           | 4096 | 36.2        | 18.3         | 1.4     |
|        | complex64 | 8    | 74.9        | 20.6         | 3.6     |
|        |           | 4096 | 74.1        | 20.8         | 3.6     |

[ghstack-poisoned]

peterbell10 committed:
Update on "Add specialized lerp kernel to accelerate `make_tensor`"


`make_tensor` used several times for every single opinfo test, so small gains in
performace will result in significant improvements in test completion time.
Aside from the random number generation, the main limitation for floating an
complex dtypes is the scaling operation from [0, 1] to [low, high]. This
operation is a linear interpolation (`lerp`) but calling `torch.lerp` is
actually not faster because the kernel isn't optimised for scalar `low` and
`high` values. Particularly on CUDA, this means extra kernel launches and device
synchronization to copy the values into tensors.

This adds a new `_lerp_scalar` kernel for use in `make_tensor` and also reduces
the number of seperate kernels launched for complex tensors, since for the small
tensors used in tests it is often the kernel launch overhead that is the
limiting factor.

Here are benchmark results for `make_tensor` with small and medium sized tensors.

| Device | dtype     | size | Master (us) | This PR (us) | Speedup |
|--------|-----------|------|-------------|--------------|---------|
| cpu    | float32   | 8    | 19.6        | 8.61         | 2.3     |
|        |           | 4096 | 34.9        | 23.3         | 1.5     |
|        | complex64 | 8    | 37.4        | 10.1         | 2.7     |
|        |           | 4096 | 70.3        | 39.2         | 1.8     |
| cuda   | float32   | 8    | 38.7        | 28.2         | 1.4     |
|        |           | 4096 | 36.2        | 18.3         | 1.4     |
|        | complex64 | 8    | 74.9        | 20.6         | 3.6     |
|        |           | 4096 | 74.1        | 20.8         | 3.6     |

[ghstack-poisoned]

vadimkantorov commented:
Unrelated lerp feature request on copycast-less support of uint8 inputs, but maybe if lerp is being updated now - a good time for this too: https://github.com/pytorch/pytorch/issues/54389

vadimkantorov commented:
Does this fix https://github.com/pytorch/pytorch/issues/57947? Does this PR support mixed scalar/tensor lo/hi?

peterbell10 committed:
Update on "Add specialized lerp kernel to accelerate `make_tensor`"


`make_tensor` used several times for every single opinfo test, so small gains in
performace will result in significant improvements in test completion time.
Aside from the random number generation, the main limitation for floating an
complex dtypes is the scaling operation from [0, 1] to [low, high]. This
operation is a linear interpolation (`lerp`) but calling `torch.lerp` is
actually not faster because the kernel isn't optimised for scalar `low` and
`high` values. Particularly on CUDA, this means extra kernel launches and device
synchronization to copy the values into tensors.

This adds a new `_lerp_scalar` kernel for use in `make_tensor` and also reduces
the number of seperate kernels launched for complex tensors, since for the small
tensors used in tests it is often the kernel launch overhead that is the
limiting factor.

Here are benchmark results for `make_tensor` with small and medium sized tensors.

| Device | dtype     | size | Master (us) | This PR (us) | Speedup |
|--------|-----------|------|-------------|--------------|---------|
| cpu    | float32   | 8    | 19.6        | 8.61         | 2.3     |
|        |           | 4096 | 34.9        | 23.3         | 1.5     |
|        | complex64 | 8    | 37.4        | 13.4         | 2.8     |
|        |           | 4096 | 70.3        | 46.0         | 1.5     |
| cuda   | float32   | 8    | 38.7        | 28.2         | 1.4     |
|        |           | 4096 | 36.2        | 18.3         | 1.4     |
|        | complex64 | 8    | 74.9        | 32.3         | 2.3     |
|        |           | 4096 | 74.1        | 32.6         | 2.3     |

[ghstack-poisoned]

peterbell10 committed:
Update on "Add specialized lerp kernel to accelerate `make_tensor`"


`make_tensor` used several times for every single opinfo test, so small gains in
performace will result in significant improvements in test completion time.
Aside from the random number generation, the main limitation for floating an
complex dtypes is the scaling operation from [0, 1] to [low, high]. This
operation is a linear interpolation (`lerp`) but calling `torch.lerp` is
actually not faster because the kernel isn't optimised for scalar `low` and
`high` values. Particularly on CUDA, this means extra kernel launches and device
synchronization to copy the values into tensors.

This adds a new `_lerp_scalar` kernel for use in `make_tensor` and also reduces
the number of seperate kernels launched for complex tensors, since for the small
tensors used in tests it is often the kernel launch overhead that is the
limiting factor.

Here are benchmark results for `make_tensor` with small and medium sized tensors.

| Device | dtype     | size | Master (us) | This PR (us) | Speedup |
|--------|-----------|------|-------------|--------------|---------|
| cpu    | float32   | 8    | 19.6        | 8.61         | 2.3     |
|        |           | 4096 | 34.9        | 23.3         | 1.5     |
|        | complex64 | 8    | 37.4        | 13.4         | 2.8     |
|        |           | 4096 | 70.3        | 46.0         | 1.5     |
| cuda   | float32   | 8    | 38.7        | 28.2         | 1.4     |
|        |           | 4096 | 36.2        | 18.3         | 1.4     |
|        | complex64 | 8    | 74.9        | 32.3         | 2.3     |
|        |           | 4096 | 74.1        | 32.6         | 2.3     |

[ghstack-poisoned]

peterbell10 committed:
Update on "Add specialized lerp kernel to accelerate `make_tensor`"


`make_tensor` used several times for every single opinfo test, so small gains in
performace will result in significant improvements in test completion time.
Aside from the random number generation, the main limitation for floating an
complex dtypes is the scaling operation from [0, 1] to [low, high]. This
operation is a linear interpolation (`lerp`) but calling `torch.lerp` is
actually not faster because the kernel isn't optimised for scalar `low` and
`high` values. Particularly on CUDA, this means extra kernel launches and device
synchronization to copy the values into tensors.

This adds a new `_lerp_scalar` kernel for use in `make_tensor` and also reduces
the number of seperate kernels launched for complex tensors, since for the small
tensors used in tests it is often the kernel launch overhead that is the
limiting factor.

Here are benchmark results for `make_tensor` with small and medium sized tensors.

| Device | dtype     | size | Master (us) | This PR (us) | Speedup |
|--------|-----------|------|-------------|--------------|---------|
| cpu    | float32   | 8    | 19.6        | 8.61         | 2.3     |
|        |           | 4096 | 34.9        | 23.3         | 1.5     |
|        | complex64 | 8    | 37.4        | 13.4         | 2.8     |
|        |           | 4096 | 70.3        | 46.0         | 1.5     |
| cuda   | float32   | 8    | 38.7        | 28.2         | 1.4     |
|        |           | 4096 | 36.2        | 18.3         | 1.4     |
|        | complex64 | 8    | 74.9        | 32.3         | 2.3     |
|        |           | 4096 | 74.1        | 32.6         | 2.3     |

[ghstack-poisoned]

peterbell10 commented:
This PR doesn't actually touch the `torch.lerp` interface.
